Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp

```diff
@@ -659,97 +659,390 @@ bool TerrainSystem::GetIsHoleFromFloats(float x, float y, Sampler sampler) const
+            // exists only if all 4 points of the + sign exist.
+            terrainExists[inPosIndex] = terrainExists[inPosIndex] && exists[queryPositionIndex + querySubindex];
+        }
 
-        // This needs better logic for handling cases where some points exist and some don't, but for now we'll say that if
-        // any of the four points exist, then the terrain exists.
-        terrainExists[i] = exists[iteratorIndex] || exists[iteratorIndex + 1] || exists[iteratorIndex + 2] || exists[iteratorIndex + 3];
+        if (terrainExists[inPosIndex])
+        {
+            // We have 4 vertices that make a + sign, cross the two lines to get the normal at the center.
+            // (right - left) x (up - down)
+            normals[inPosIndex] = (queryPositions[queryPositionIndex + 2] - queryPositions[queryPositionIndex + 1])
+                                .Cross(queryPositions[queryPositionIndex + 3] - queryPositions[queryPositionIndex])
+                                .GetNormalized();
+        }
+        else
+        {
+            // If at least one of the 4 points of the + sign didn't exist, just give it a Z-up normal.
+            normals[inPosIndex] = AZ::Vector3::CreateAxisZ();
+        }
     }
 }
 
-AZ::Vector3 TerrainSystem::GetNormalSynchronous(float x, float y, Sampler sampler, bool* terrainExistsPtr) const
+void TerrainSystem::GetNormalsSynchronousClamp(
+    const AZStd::span<const AZ::Vector3>& inPositions,
+    AZStd::span<AZ::Vector3> normals,
+    AZStd::span<bool> terrainExists) const
 {
-    AZStd::shared_lock<AZStd::shared_mutex> lock(m_areaMutex);
+    // When querying for normals with a CLAMP sampler, we divide each terrain grid square into two triangles, and return the normal
+    // for the triangle that the requested position falls on.
+    // Right now, the terrain system will always split each terrain grid square like this:  |\|
+    // Eventually, the terrain system might get more complicated per-square logic, at which point the logic below would need
+    // to change to account for the per-square triangle split direction.
 
-    bool terrainExists = false;
+    const float queryResolution = m_currentSettings.m_heightQueryResolution;
 
-    AZ::Vector3 outNormal = AZ::Vector3::CreateAxisZ();
+    // The number of points we're querying for each normal
+    const size_t queryCount = 3;
 
-    if (!InWorldBounds(x, y))
+    // The full set of positions to query to be able to calculate all the normals.
+    AZStd::vector<AZ::Vector3> queryPositions;
+    queryPositions.reserve(inPositions.size() * queryCount);
+
+    // To determine if terrain exists at the queried normal, we'll use the terrain exists flag from the terrain grid point closest
+    // to the position. We'll track this while building up our query positions.
+    AZStd::vector<uint8_t> existsIndex;
+    existsIndex.reserve(inPositions.size());
+
+    for (auto& position : inPositions)
```

Comment:
Nitpick: Can this be `const auto& position` ?

Reply:
It's not technically needed - since inPositions is already a span of const Vector3, this already ends up const. I'll add it to make it more explicit and obvious though. :)